### PR TITLE
Add build matrix for docker images, update images

### LINKED ISSUE #2164: container image has no cudnn
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

The container images, without tagged "latest" (the oldest version of the image), have no libcudnn.so.8, and I cannot use GPU acceleration with these images.
It seems that these images were not built with this [Dockerfile](https://github.com/DeepLabCut/DeepLabCut/tree/main/docker), and these base images do not use "nvidia/cuda:11.X.Y-cudnn8" looking at the [history](https://hub.docker.com/layers/deeplabcut/deeplabcut/2.2.1.1-core-cuda11.0.3-runtime-ubuntu18.04/images/sha256-6d8c1f80edd8c0081e1939c31f322f2a30cebba32fdb8c92cd933e9a89bada84?context=explore).

### Operating System

operating system: Linux

### DeepLabCut version

dlc version: 2.2.1.1, 2.2.0.6

### DeepLabCut mode

single animal

### Device type

gpu: NVIDIA GeForce RTX3080

### Steps To Reproduce

`docker run --rm -it --gpus all deeplabcut/deeplabcut:2.2.1.1-gui-cuda11.7.0-runtime-ubuntu20.04 python3 -c 'from tensorflow.python.client import device_lib; device_lib.list_local_devices()'`

### Relevant log output

```shell
2023-02-25 13:41:14.467894: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0
2023-02-25 13:41:15.138019: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-02-25 13:41:15.139964: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcuda.so.1
2023-02-25 13:41:15.156153: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:923] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node
Your kernel may have been built without NUMA support.
2023-02-25 13:41:15.156207: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: NVIDIA GeForce RTX 3080 computeCapability: 8.6
coreClock: 1.71GHz coreCount: 70 deviceMemorySize: 12.00GiB deviceMemoryBandwidth: 849.46GiB/s
2023-02-25 13:41:15.156236: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0
2023-02-25 13:41:15.159535: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublas.so.11
2023-02-25 13:41:15.159597: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcublasLt.so.11
2023-02-25 13:41:15.177416: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcufft.so.10
2023-02-25 13:41:15.177711: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcurand.so.10
2023-02-25 13:41:15.178288: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcusolver.so.11
2023-02-25 13:41:15.178931: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcusparse.so.11
2023-02-25 13:41:15.179111: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2023-02-25 13:41:15.179138: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1766] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2023-02-25 13:41:15.280587: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
2023-02-25 13:41:15.280631: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      0 
2023-02-25 13:41:15.280648: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1277] 0:   N
```

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by stes**
- Improvements to the build process for docker images: With the specified build matrix, it is now easier to build images with different DeepLabCut and CUDA base container versions
- Support two version of deeplabcut, as well as CUDA 11.4 (as before) and the more recent CUDA 11.7

Open:
- [x] Build containers, and attach the full build log
- [x] Update containers on DockerHub
- [x] Discuss addition of additional CUDA versions/base images
- [ ] Update PyPI package to interface with the containers

## Build and test logs

https://gist.github.com/stes/80412bac41a37cabb6248db5e5d2884e

## Testing

If you come across this PR and want to test it, please see [this discussion](https://forum.image.sc/t/deeplabcut-docker-tensorflow-gpu-problem/70044/7?u=stes) for instructions (until this package is updated):

- Download [this experimental version of `deeplabcut-docker`](https://gist.github.com/stes/4dfb0c7e95a3ac3b2356551e8d375254)
- Run the following:
```
# make executable
chmod +x ./deeplabcut_docker.sh
# pull the image
docker pull deeplabcut/deeplabcut:2.2.1.1-gui-cuda11.7.0-runtime-ubuntu20.04
# run the UI
./deeplabcut_docker.sh gui
```

This should open the deeplabcut UI. If you are on a linux machine, you might run `xhost +` before executing the last command if you get a permission error.

Fixes #2164

### CODE REVIEW

File: docker/Dockerfile.base
Comment (natsukium) [inline docker/Dockerfile.base]: It seems that `CUDA_VERSION` should be `11.4.0` only because the latter info is not the cuda version.

File: docker/Dockerfile.base
Comment (natsukium) [inline docker/Dockerfile.base:2]: Since this image does not contain cudnn8, could you please use nvidia/cuda:${CUDA_VERSION}-cudnn8-runtime-ubuntu20.04?

File: docker/Dockerfile.core
Comment (natsukium) [inline docker/Dockerfile.core]: ditto

File: docker/Dockerfile.gui
Comment (natsukium) [inline docker/Dockerfile.gui]: ditto

File: docker/Dockerfile.jupyter
Comment (natsukium) [inline docker/Dockerfile.jupyter]: ditto

File: docker/Dockerfile.test
Comment (natsukium) [inline docker/Dockerfile.test]: ditto

File: docker/build.sh
Comment (natsukium) [inline docker/build.sh]: ditto

File: docker/build.sh
Comment (natsukium) [inline docker/build.sh]: The tag also needs to be changed.

**Comment by MMathisLab** [comment]
@stes @jeylau let's finish and merge

**Comment by MMathisLab** [comment]
@jeylau can we fix this? // get it merged DLC broader docker support is quite important for a lot of applications, including in our lab ;)

File: docker/Dockerfile.base
Comment (jeylau) [inline docker/Dockerfile.base]:
```suggestion
ARG CUDA_VERSION=11.4.0-cudnn8-runtime-ubuntu20.04
```

File: docker/Dockerfile.core
Comment (jeylau) [inline docker/Dockerfile.core]:
```suggestion
ARG CUDA_VERSION=11.4.0-cudnn8-runtime-ubuntu20.04
```

File: docker/Dockerfile.gui
Comment (jeylau) [inline docker/Dockerfile.gui]:
```suggestion
ARG CUDA_VERSION=11.4.0-cudnn8-runtime-ubuntu20.04
```

File: docker/Dockerfile.jupyter
Comment (jeylau) [inline docker/Dockerfile.jupyter]:
```suggestion
ARG CUDA_VERSION=11.4.0-cudnn8-runtime-ubuntu20.04
```

File: docker/Dockerfile.test
Comment (jeylau) [inline docker/Dockerfile.test]:
```suggestion
ARG CUDA_VERSION=11.4.0-cudnn8-runtime-ubuntu20.04
```

File: docker/build.sh
Comment (jeylau) [inline docker/build.sh]:
```suggestion
        11.4.0-cudnn8-runtime-ubuntu20.04 \
```

File: docker/build.sh
Comment (jeylau) [inline docker/build.sh]:
```suggestion
        11.7.0-cudnn8-runtime-ubuntu20.04
```

**Comment by jeylau** [comment]
"core" now works; I will test the GUI tomorrow on Ubuntu.

**Review by maryapp — APPROVED** [review]

**Comment by maryapp** [comment]
@MMathisLab, can you please, merge this branch? 

- The matrix code is functional
- the Dockerfile was adjusted by Jessy (it's minor changes dor cuda & cudnn, but in any cases ARG parameters come from matrix bash script, so it's mostly for coherency). 

Once it merged, I will PR to:
- [ ] : adjust docker developer-guide README
- [ ] the change versions in Dockerfile & matrix that will correspond to images on dockerhub. 
(I can't PR beautifully now, as it's fork from external branch, and I would prefer to avoid new forks/PR on stes repo, to keep the history more linear)

**Review by n-poulsen — APPROVED** [review]

**Comment by MMathisLab** [comment]
@maryapp can we please finalize this? cc @stes

**Comment by maryapp** [comment]
@MMathisLab Here is my [PR ](https://github.com/stes/DeepLabCut/pull/3) to @stes PR! Hope this time it will make the affair! 🙌

**Comment by stes** [comment]
All right, some progress here!

Thanks a lot to @maryapp for taking over and adding support for the latest deeplabcut version / upgrading the images.

The changelog can be seen in my forked repo here: https://github.com/stes/DeepLabCut/pull/3

If all tests pass (for docker --- I assume @maryapp checked these), this is good to go for merging into main from my end.

cc @MMathisLab @AlexEMG

**Comment by stes** [comment]
An open Q is maybe the faith of the [deeplabcut-docker](https://pypi.org/project/deeplabcut-docker/) python package. It seems GUI support will be dropped with the latest version of this PR --- so I think we should decide whether to remove that, or use the legacy images in the pypi package. Opinions?

**Comment by maryapp** [comment]
> If all tests pass (for docker --- I assume @maryapp checked these), this is good to go for merging into main from my end.

Yes, all passed.

> An open Q is maybe the faith of the [deeplabcut-docker](https://pypi.org/project/deeplabcut-docker/) python package. It seems GUI support will be dropped with the latest version of this PR --- so I think we should decide whether to remove that, or use the legacy images in the pypi package. Opinions?

I would vote for the 1st option. I will check on the python package code (also how it picks up the right cuda version support, etc.)

**Comment by MMathisLab** [comment]
>> use the legacy images in the pypi package. Opinions?

We need to keep the GUI support in legacy versions. 

Please then decide if this can be merged ...

**Comment by maryapp** [comment]
@stes can you, please, elaborate what is "legacy" images for you? You are referencing the very first old images that are distributed via docker pip package now because of the tag?

**Comment by MMathisLab** [comment]
I am merging as this is becoming unmanageable -- we need docker support...

**Pull request merged**
